### PR TITLE
Added Example

### PR DESCRIPTION
**Pull request by kushagra414**
Added an example showing how the Component transmit and receive data from other Components

### CODE REVIEW

**Comment by kushagra414** [comment]
@orensbruli Please Review

File: getting-started/python/interaction/implements/src/specificworker.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+		# The API of python-innermodel is not exactly the same as the C++ version
+		# self.innermodel.updateTransformValues("head_rot_tilt_pose", 0, 0, 0, 1.3, 0, 0)
+		# z = librobocomp_qmat.QVec(3,0)
+		# r = self.innermodel.transform("rgbd", z, "laser")
+		# r.printvector("d")
+		# print r[0], r[1], r[2]
+		self.image = cv2.imread(self.path)
+		self.width,self.height,self.depth = self.image.shape
+		self.image = np.array(self.image,dtype=np.uint8)
+		return True
```
Comment (orensbruli) [inline getting-started/python/interaction/implements/src/specificworker.py]: It's reading the same image every time the compute slot is called.

File: getting-started/python/interaction/implements/README.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+- Generate a Component using the CDSL file.<br>
+`robocompdsl ImplementCameraSimple.cdsl .`
+
+- Verify that you have installed `cv2` and `numpy`<br>
+`pip install cv2`<br>
+`pip install numpy`
+
+- Open the `CameraSimple.idsl` file in `/opt/robocomp/interfaces/IDSLs` directory. As you can see, there is only one function i.e. `getImage()`. This function outputs a structure called `TImage` and the structure itself has 4 components which stores  of height, width, depth, and Image itself.
+
+- Remember the Ice middleware only supports data type which are python native data structure, hence you when you transmit data type such as numpy `arrays` or pandas `data frame`. They will be converted to string.
```
Comment (orensbruli) [inline getting-started/python/interaction/implements/README.md]: Ice is not related to python this way. Ice has its own types and the documentation for the IDSL should be read for this.
Are you sure that those specific types are converted to a string?

File: getting-started/python/interaction/requires/README.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
+# This directory contains the implementation of RGB component.
+
+After implementing the RGB component. We can create another component that will use the information transmitted by the RGB component.
+
+This folder contains only those files which need to be changed to implement a RGB component i.e. the `CDSL` file, the specificworker.py file in src/ directory, and the config file in etc/ directory.
+
+- If you open the `CSDL` file, you can see `requires` keyword is used, which specifies that the component is required here.
```
Comment (orensbruli) [inline getting-started/python/interaction/requires/README.md]: **requires** reference is to the interface, not for a component.

**Review by orensbruli — COMMENTED** [review]
I would need my workmates to review this also. I think it's a little bit confusing example. I understand you want to illustrate the communication between two components through an interface that one requires and the other implements. Maybe it would be better to read from a real camera or at least a video because it can be confusing to see always the same image. Also, binary files would be very limited in the repository (one image is ok, for this example, but I would not upload the video if we choose to do that way).

File: getting-started/python/interaction/implements/src/specificworker.py (hunk 1 above)
Comment (kushagra414) [reply getting-started/python/interaction/implements/src/specificworker.py]: I update the code to read from a webcam.

File: getting-started/python/interaction/implements/README.md (hunk 2 above)
Comment (kushagra414) [reply getting-started/python/interaction/implements/README.md]: Slice maps a sequence of bytes to string in python2 by default. I fixed it and gave reference to the [website](https://doc.zeroc.com/ice/3.7/language-mappings/python-mapping/client-side-slice-to-python-mapping)

File: getting-started/python/interaction/requires/README.md (hunk 3 above)
Comment (kushagra414) [reply getting-started/python/interaction/requires/README.md]: Fixed

**Comment by kushagra414** [comment]
I fixed the issues, Thanks for pointing out the mapping problem. Now I know better how the mapping works in Slice to python.

**Review by pilarbachiller — CHANGES_REQUESTED** [review]
You should try to explain the example a little more. It is not clear what is the main objective. Maybe the name "interaction" is not proper. One could think it is related to HRI or something like that. Also, if you want to explain again what is Ice, although maybe it is unnecessary, you can search for a more formal definition.

**Comment by kushagra414** [comment]
> You should try to explain the example a little more. It is not clear what is the main objective. Maybe the name "interaction" is not proper. One could think it is related to HRI or something like that. Also, if you want to explain again what is Ice, although maybe it is unnecessary, you can search for a more formal definition.

@pilarbachiller Is this fine?
I think one should know what is ICE, objects, proxies, servers, and the client to fully grasp the working of robocomp.

**Comment by pbustos** [comment]
Please use an existing interface from the ~/robocomp/interfaces folder.

**Comment by kushagra414** [comment]
> Please use an existing interface from the ~/robocomp/interfaces folder.

@pbustos I have used the CameraSimple interface which exists in the interface directory

File: getting-started/python/Client and Server Structure/README.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
+# README
+
+## The Internet Communications Engine(ICE)
+- Ice is an object-oriented middleware platform. Fundamentally, this means that Ice provides tools, APIs, and library support for building object-oriented client-server applications. Ice applications are suitable for use in heterogeneous environments: client and server can be written in different programming languages, can run on different operating systems and machine architectures, and can communicate using a variety of networking technologies. The source code for these applications is portable regardless of the deployment environment.
```
Comment (orensbruli) [inline getting-started/python/Client and Server Structure/README.md]: https://doc.zeroc.com/ice/3.7/ice-overview/ice-architecture

File: getting-started/python/Client and Server Structure/README.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    - Clients are active entities. They issue requests for service to servers.
+    - Servers are passive entities. They provide services in response to client requests.
+- Frequently, servers are not "pure" servers, in the sense that they never issue requests and only respond to requests. Instead, servers often act as a server on behalf of some client but, in turn, act as a client to another server in order to satisfy their client's request.
+- Similarly, clients often are not "pure" clients, in the sense that they only request service from an object. Instead, clients are frequently client-server hybrids. For example, a client might start a long-running operation on a server; as part of starting the operation, the client can provide a callback object to the server that is used by the server to notify the client when the operation is complete. In that case, the client acts as a client when it starts the operation, and as a server when it is notified that the operation is complete.
+
+- Such role reversal is common in many systems, so, frequently, client-server systems could be more accurately described as peer-to-peer systems.
```
Comment (orensbruli) [inline getting-started/python/Client and Server Structure/README.md]: https://doc.zeroc.com/technical-articles/general-topics/chat-demo/concepts#Concepts-ClientsandServers

File: getting-started/python/Client and Server Structure/README.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+## Objects & Proxies
+
+- An Ice object is an entity in the local or a remote address space that can respond to client requests. The objects are created by Server
+- Each Ice object has a unique object identity. An object's identity is an identifying value that distinguishes the object from all other objects. The Ice object model assumes that object identities are globally unique, that is, no two objects within an Ice communication domain can have the same object identity.
```
Comment (orensbruli) [inline getting-started/python/Client and Server Structure/README.md]: https://doc.zeroc.com/ice/3.7/ice-overview/ice-architecture/terminology#Terminology-IceObjects

File: getting-started/python/Client and Server Structure/README.md [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+- For a client to be able to contact an Ice object, the client must hold a proxy for the Ice object. A proxy is an artifact that is local to the client's address space; it represents the (possibly remote) Ice object for the client. A proxy acts as the local ambassador for an Ice object: when the client invokes an operation on the proxy, the Ice run time:
+    1. Locates the Ice object
+    2. Activates the Ice object's server if it is not running
+    3. Activates the Ice object within the server
+    4. Transmits any in-parameters to the Ice object
+    5. Waits for the operation to complete
+    6. Returns any out-parameters and the return value to the client (or throws an exception in case of an error)
+- A proxy encapsulates all the necessary information for this sequence of steps to take place.
```
Comment (orensbruli) [inline getting-started/python/Client and Server Structure/README.md]: https://doc.zeroc.com/ice/3.7/ice-overview/ice-architecture/terminology#Terminology-Proxies

File: getting-started/python/Client and Server Structure/README.md [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+- You can find the object address in the `implements/etc/config`.
+- You can find the proxy address in the `requires/etc/config`.
```
Comment (orensbruli) [inline getting-started/python/Client and Server Structure/README.md]: I don't know what those paths mean.

Comment (orensbruli) [reply getting-started/python/Client and Server Structure/README.md]: Oh, are paths to your own example, right? Maybe a link to the files would help to understand it.

File: getting-started/python/Client and Server Structure/README.md [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+- In this tutorial, We are going to learn how the different components in robocomp, share data using the Ice's peer-to-peer systems. 
+- For the data to be shared between components, we need to have at least 2 components i.e. One component will act as a server and the other will be the client.
+
+- We can use `implement` or `publish` keyword in the `Component Specific Definition Language(.cdsl)` to specify that the component will be implemented and generally this component will generate some kind of data that might be required by another component of robocomp.
+
+- We can use `require` or `subscribe` keywords in `Component Specific Definition Language(.cdsl)` to specify that the current component needs some information from the required component of the robocomp.
+
+- We will create a server and a client for CameraSimple component that will show you how you can share an Image data(Information) from client to server.
+
+- The implements folder contains implementation of the CameraSimple(will act as a Server). And the requires folder contains a component that requires the information send by CameraSimple component.
```
Comment (orensbruli) [inline getting-started/python/Client and Server Structure/README.md]: And maybe a pair of links to the Readmes of the implements and requires folders to be able to follow the tutorial?

**Comment by orensbruli** [comment]
@kushagra414 Usually it's not a good idea to copy and paste the exact same text that you find in other online documentation. I understand the intention and I think it's a good idea, but I think it would be better to remove the paragraphs and link to those specific sections of the zeroc documentation with a little introduction telling the user what they should read and why.

**Comment by kushagra414** [comment]
> @kushagra414 Usually it's not a good idea to copy and paste the exact same text that you find in other online documentation. I understand the intention and I think it's a good idea, but I think it would be better to remove the paragraphs and link to those specific sections of the zeroc documentation with a little introduction telling the user what they should read and why.

Ok, I will update the PR immediately Thanks. I did because @pilarbachiller asked me to write some formal definitions. Anyways I will keep this in mind next time.

**Comment by kushagra414** [comment]
@orensbruli I updated the README. What do you think? Is this fine or do I need to change something else?

File: getting-started/python/Client and Server Structure/README.md [hunk 10]
```diff
@@ -0,0 +1,36 @@
+# README
+
+To understand how the robocomp works internally one should know answers to questions like how the robocomp's components share data, how data structure from one language is mapped to the other, What are proxies, server, etc.
```
Comment (orensbruli) [inline getting-started/python/Client and Server Structure/README.md]:
```suggestion
To understand how the robocomp works internally one should know answers to questions like how the robocomp's components share data, how data structure from one language is mapped to the other, what are proxies, server, etc.
```

File: getting-started/python/Client and Server Structure/README.md [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+## The Internet Communications Engine(ICE)
+- Robocomp uses The Internet Communications Engine(ICE) which helps the components written in different languages communicate.
+- Ice is an object-oriented middleware platform. Which helps programs communicate by creating servers and clients. To know more about ICE you can read the ICE [documentation](https://doc.zeroc.com/ice/3.7/ice-overview/ice-architecture). 
+
+## Servers & Clients
+- An application will act as a server if it transmits some kind of information.
+- An application that requests' server for information is called the client.
+- Applications are not necessarily servers or clients but generally mix of those two i.e. peers.
+- Servers and Clients are used by robocomp's component to share data.
+- You should read more about ICE servers, Clients and architecture You can find the documentation [here](https://doc.zeroc.com/technical-articles/general-topics/chat-demo/concepts#Concepts-ClientsandServers).
```
Comment (orensbruli) [inline getting-started/python/Client and Server Structure/README.md]:
```suggestion
- You should read more about ICE servers, clients and architecture. You can find the documentation [here](https://doc.zeroc.com/technical-articles/general-topics/chat-demo/concepts#Concepts-ClientsandServers). 
```

**Review by orensbruli — COMMENTED** [review]
Check this minor typos.

**Comment by kushagra414** [comment]
> Check this minor typos.

@orensbruli Done.

**Comment by orensbruli** [comment]
Thank you so much for your time and effort @kushagra414

**Pull request merged**